From 72f35948813343b70af589403b4f9ca3cbfdc18e Mon Sep 17 00:00:00 2001
From: "jiaheng.fan" <jiaheng.fan@nxp.com>
Date: Thu, 25 May 2017 11:16:15 +0800
Subject: [PATCH 051/741] dmaengine: nxp-dpaa2-qdma: support qDMA from LS2 rev2

commit 4b8ee0278e39e1e8db71d5b831ae70c61b6df8e3 from
http://source.codeaurora.org/external/qoriq/qoriq-components/linux

Update qDMA driver to work with qDMA IP ver 1.1 from LS2 rev2.
The driver is still compatible with IP v1.0 from rev1.
Set the transaction type for command field for both source and
destination descriptors to coherent copy of cacheable memory.
Don't enable source prefetchable (SPF) in IP v1.0 FD since it
overlaps with IP v1.1 source read transaction QoS field.

Signed-off-by: Catalin Horghidan <catalin.horghidan@nxp.com>
Signed-off-by: jiaheng.fan <jiaheng.fan@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/dma/dpaa2-qdma/dpaa2-qdma.c | 7 ++++---
 drivers/dma/dpaa2-qdma/dpaa2-qdma.h | 7 ++++++-
 2 files changed, 10 insertions(+), 4 deletions(-)

diff --git a/drivers/dma/dpaa2-qdma/dpaa2-qdma.c b/drivers/dma/dpaa2-qdma/dpaa2-qdma.c
index e0f35e54232b..53842761c4d4 100644
--- a/drivers/dma/dpaa2-qdma/dpaa2-qdma.c
+++ b/drivers/dma/dpaa2-qdma/dpaa2-qdma.c
@@ -129,7 +129,7 @@ static void dpaa2_qdma_populate_fd(uint32_t format,
 	fd->simple.bpid = QMAN_FD_BMT_ENABLE;
 	fd->simple.format_offset = QMAN_FD_FMT_ENABLE | QMAN_FD_SL_DISABLE;
 
-	fd->simple.frc = format | QDMA_SER_CTX | QDMA_FD_SPF_ENALBE;
+	fd->simple.frc = format | QDMA_SER_CTX;
 	fd->simple.ctrl = QMAN_FD_VA_DISABLE |
 			QMAN_FD_CBMT_ENABLE | QMAN_FD_SC_DISABLE;
 }
@@ -142,10 +142,11 @@ static void dpaa2_qdma_populate_first_framel(
 	struct dpaa2_qdma_sd_d *sdd;
 
 	sdd = (struct dpaa2_qdma_sd_d *)dpaa2_comp->desc_virt_addr;
+	memset(sdd, 0, 2 * (sizeof(*sdd)));
 	/* source and destination descriptor */
-	sdd->cmd = CMD_TTYPE_RW; /* source descriptor CMD */
+	sdd->cmd = QDMA_SD_CMD_RDTTYPE_COHERENT; /* source descriptor CMD */
 	sdd++;
-	sdd->cmd = CMD_TTYPE_RW; /* destination descriptor CMD */
+	sdd->cmd = QDMA_DD_CMD_WRTTYPE_COHERENT; /* dest descriptor CMD */
 
 	memset(f_list, 0, sizeof(struct dpaa2_frame_list));
 	/* first frame list to source descriptor */
diff --git a/drivers/dma/dpaa2-qdma/dpaa2-qdma.h b/drivers/dma/dpaa2-qdma/dpaa2-qdma.h
index f692858c2524..1f3380c7da7c 100644
--- a/drivers/dma/dpaa2-qdma/dpaa2-qdma.h
+++ b/drivers/dma/dpaa2-qdma/dpaa2-qdma.h
@@ -61,7 +61,12 @@ struct dpaa2_qdma_sd_d {
 	uint32_t rbpcmd;	/* Route-by-port command */
 	uint32_t cmd;
 } __attribute__((__packed__));
-#define CMD_TTYPE_RW (0x4 << 28)
+/* Source descriptor command read transaction type for RBP=0:
+ coherent copy of cacheable memory */
+#define QDMA_SD_CMD_RDTTYPE_COHERENT (0xb << 28)
+/* Destination descriptor command write transaction type for RBP=0:
+ coherent copy of cacheable memory */
+#define QDMA_DD_CMD_WRTTYPE_COHERENT (0x6 << 28)
 
 #define QDMA_SG_FMT_SDB		0x0 /* single data buffer */
 #define QDMA_SG_FMT_FDS		0x1 /* frame data section */
-- 
2.17.1

